extensions/typescript-language-features/src/typescriptServiceClient.ts

```diff
@@ -973,6 +985,64 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			case EventName.projectLoadingFinish:
 				this.loadingIndicator.finishedLoadingProject((event as Proto.ProjectLoadingFinishEvent).body.projectName);
 				break;
+
+			case EventName.createDirectoryWatcher:
+				this.createFileSystemWatcher(
+					(event.body as Proto.CreateDirectoryWatcherEventBody).id,
+					new vscode.RelativePattern(
+						vscode.Uri.file((event.body as Proto.CreateDirectoryWatcherEventBody).path),
+						(event.body as Proto.CreateDirectoryWatcherEventBody).recursive ? '**' : '*'
+					)
+				);
+				break;
+
+			case EventName.createFileWatcher:
+				this.createFileSystemWatcher(
+					(event.body as Proto.CreateFileWatcherEventBody).id,
+					new vscode.RelativePattern(
+						vscode.Uri.file((event.body as Proto.CreateFileWatcherEventBody).path),
+						'*'
+					)
+				);
+				break;
+
+			case EventName.closeFileWatcher:
+				this.closeFileSystemWatcher(event.body.id);
+				break;
+		}
+	}
+
+	private createFileSystemWatcher(
+		id: number,
+		pattern: vscode.RelativePattern
+	) {
+		const disposable = new DisposableStore();
+
+		const watcher = disposable.add(vscode.workspace.createFileSystemWatcher(pattern, { excludes: [] /* TODO need to fill in excludes list */ }));
+
+		disposable.add(watcher.onDidChange(changeFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: changeFile.fsPath, eventType: 'update' } satisfies Proto.WatchChangeRequestArgs)
+		));
+		disposable.add(watcher.onDidCreate(createFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: createFile.fsPath, eventType: 'create' } satisfies Proto.WatchChangeRequestArgs)
+		));
+		disposable.add(watcher.onDidDelete(deletedFile =>
+			this.executeWithoutWaitingForResponse('watchChange', { id, path: deletedFile.fsPath, eventType: 'delete' } satisfies Proto.WatchChangeRequestArgs)
+		));
```

Comment:
@sheetalkamat I wonder if the communication between client and server should work in a way that we can collect multiple file events and send them over in batches. Here we seem to be doing 1 IPC roundtrip per single file event, but in reality there will be hundreds of events. 

Maybe that could go together with a utility to aggregate over multiple events for when we deal with changes to `node_modules` where you are only interested in the top level module event.